types/bluebird/bluebird-tests.ts

```diff
@@ -373,6 +373,30 @@ fooProm = fooProm.catch(CustomError, reason => {
 	fooProm = fooProm.catch(CustomError1, CustomError2, CustomError3, error => {});
 	fooProm = fooProm.catch(CustomError1, CustomError2, CustomError3, CustomError4, error => {});
 	fooProm = fooProm.catch(CustomError1, CustomError2, CustomError3, CustomError4, CustomError5, error => {});
+
+	const booPredicate1 = (error: CustomError1) => true;
+	const booPredicate2 = (error: [number]) => true;
+	const booPredicate3 = (error: string) => true;
+	const booPredicate4 = (error: Object) => true;
+	const booPredicate5 = (error: any) => true;
+
+	fooProm = fooProm.catch(booPredicate1, error => {});
+	fooProm = fooProm.catch(booPredicate1, booPredicate2, error => {});
+	fooProm = fooProm.catch(booPredicate1, booPredicate2, booPredicate3, error => {});
+	fooProm = fooProm.catch(booPredicate1, booPredicate2, booPredicate3, booPredicate4, error => {});
+	fooProm = fooProm.catch(booPredicate1, booPredicate2, booPredicate3, booPredicate4, booPredicate5, error => {});
+
+	const booObject1 = new CustomError1();
+	const booObject2 = [400, 500];
+	const booObject3 = ["Error1", "Error2"];
+	const booObject4 = {code: 400};
+	const booObject5: any = null;
+
+	fooProm = fooProm.catch(booObject1, error => {});
+	fooProm = fooProm.catch(booObject1, booObject2, error => {});
+	fooProm = fooProm.catch(booObject1, booObject2, booObject3, error => {});
+	fooProm = fooProm.catch(booObject1, booObject2, booObject3, booObject4, error => {});
+	fooProm = fooProm.catch(booObject1, booObject2, booObject3, booObject4, booObject5, error => {});
```

Comment:
Well, the changes I did for this PR were mostly in this way

To check for `Error` instances

```diff
catch<E1 extends Error, ..., E5 extends Error>
-  filter1: (new (...args: any[]) => E1) | ((error: E1) => boolean) | E1,
+  filter1: (new (...args: any[]) => E1),
```

and for any primitive type (as bluebird allow it)

```diff
catch<E1, ..., E5>(
-  filter1: ((error: E1) => boolean) | E1,
+  filter1: ((error: E1) => boolean) | E1 & object,
```

But, the problem remains in that `E1` is (still) too generic, so I'm afraid you may end up loosing type check for error instances.

The only thing that comes to my mind is to have a _complement_ type, which excludes all possible values between 2 types, something like `E excepts Error` or `E excludes Error`:

```typescript
catch<E1 extends Error, ..., E5 extends Error>
  filter1: (new (...args: any[]) => E1) | ((error: E1) => boolean) | E1,
```

```typescript
catch<E1 excepts Error, ..., E5 excepts Error>(
  filter1: ((error: E1) => boolean) | E1,
```